File: jornada/usb_doctor.py
"""The USB doctor: which USB device should carry the link, and why or why not.

Combines the I/O Registry view (:mod:`usb_registry`), the driver table
(:mod:`usb_profiles`), the optional pinned port (``~/.jornada-link/serial``)
and what is known about the handheld into one :class:`Diagnosis` — a ranked
list of serial-port candidates, the recommended port, and human findings.

The ranking is deliberately simple and deterministic: a pinned port that
exists wins; otherwise a bridge built into the dock beats a plain adapter; a
node owned by Apple's built-in driver beats a vendor extension's node (the
latter may not be openable by the user, see the ``writable`` flag); and a node
this user can actually open beats one it cannot.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

from .usb_profiles import (ARCH_ARM, ARCH_SH3, DRIVER_BUILT_IN, ROLE_DOCK_BRIDGE, ROLE_SERIAL_BRIDGE,
                           ROLE_WINCE_USB_SYNC, Classification, HandheldModel, classify,
                           identify_handheld)
from .usb_registry import UsbDevice, usb_devices

LEVEL_OK = "ok"
LEVEL_INFO = "info"
LEVEL_WARN = "warn"
LEVEL_ERROR = "error"

PIN_FILE_NAME = "serial"
DOCS_PATH = "docs/usb-link.md"

# Same rule as bin/jornada-ppp and PppController.isValidSerialPath: the path
# ends up in a root pppd command line, so only plain /dev/cu.* nodes qualify.
_SERIAL_PATH_RE = re.compile(r"^/dev/cu\.[A-Za-z0-9._-]+$")


@dataclass(frozen=True)
class Finding:
    level: str
    title: str
    detail: str


@dataclass(frozen=True)
class ClassifiedDevice:
    device: UsbDevice
    classification: Optional[Classification]

    @property
    def role(self) -> Optional[str]:
        return self.classification.role if self.classification else None


@dataclass(frozen=True)
class Candidate:
    path: str
    driver: str
    writable: bool
    device: UsbDevice
    classification: Classification
    score: int


@dataclass(frozen=True)
class Diagnosis:
    devices: Tuple[ClassifiedDevice, ...]
    candidates: Tuple[Candidate, ...]
    recommended: Optional[str]
    pinned: Optional[str]
    handheld: Optional[HandheldModel]
    findings: Tuple[Finding, ...]

    @property
    def worst_level(self) -> str:
        order = (LEVEL_ERROR, LEVEL_WARN, LEVEL_INFO, LEVEL_OK)
        return next((level for level in order if any(f.level == level for f in self.findings)), LEVEL_OK)


def is_valid_serial_path(path: str) -> bool:
    return bool(_SERIAL_PATH_RE.match(path))


def default_pin_path() -> Path:
    return Path.home() / ".jornada-link" / PIN_FILE_NAME


def read_pin(path: Optional[Path] = None) -> Optional[str]:
    """The pinned serial node, or None when unset or invalid."""
    try:
        text = (path or default_pin_path()).read_text(encoding="utf-8").strip().splitlines()
    except (OSError, ValueError):
        return None
    first = text[0].strip() if text else ""
    return first if is_valid_serial_path(first) else None


def write_pin(device_path: str, path: Optional[Path] = None) -> Path:
    if not is_valid_serial_path(device_path):
        raise ValueError(f"not a plain /dev/cu.* node: {device_path!r}")
    target = path or default_pin_path()
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(device_path + "\n", encoding="utf-8")
    return target


def clear_pin(path: Optional[Path] = None) -> bool:
    target = path or default_pin_path()
    try:
        target.unlink()
    except FileNotFoundError:
        return False
    return True


def classify_devices(devices: Tuple[UsbDevice, ...]) -> Tuple[ClassifiedDevice, ...]:
    return tuple(
        ClassifiedDevice(d, classify(d.vendor_id, d.product_id, d.product, d.serial, d.interface_classes))
        for d in usb_devices(devices))


def _score(candidate_role: str, driver: str, writable: bool, pinned: bool) -> int:
    score = {ROLE_DOCK_BRIDGE: 400, ROLE_SERIAL_BRIDGE: 200}.get(candidate_role, 0)
    if driver.startswith("com.apple."):
        score += 100
    score += 50 if writable else -500
    if pinned:
        score += 10_000
    return score


def rank_candidates(devices: Tuple[UsbDevice, ...], pinned: Optional[str] = None) -> Tuple[Candidate, ...]:
    """Every serial node of every recognised bridge, best first."""
    found: List[Candidate] = []
    for item in classify_devices(devices):
        if item.classification is None or item.role == ROLE_WINCE_USB_SYNC:
            continue
        for node in item.device.serial_nodes:
            found.append(Candidate(
                path=node.path, driver=node.driver, writable=node.writable, device=item.device,
                classification=item.classification,
                score=_score(item.role or "", node.driver, node.writable, pinned == node.path)))
    return tuple(sorted(found, key=lambda c: (-c.score, c.path)))


def recommend_serial_path(devices: Tuple[UsbDevice, ...], pinned: Optional[str] = None) -> Optional[str]:
    ranked = rank_candidates(devices, pinned)
    return ranked[0].path if ranked else None


def handheld_from_state(state: Optional[Dict[str, Any]]) -> Optional[HandheldModel]:
    """The handheld the dccm listener last recorded (``~/.jornada-link/connection.json``)."""
    if not state:
        return None
    device = state.get("device")
    if not isinstance(device, dict):
        return None
    hardware = device.get("hardware")
    name = device.get("name")
    return identify_handheld(hardware if isinstance(hardware, str) else None,
                             name if isinstance(name, str) else None)


def _dock_findings(handheld: Optional[HandheldModel], ranked: Tuple[Candidate, ...]) -> List[Finding]:
    if handheld is None:
        return []
    dock_bridge = next((c for c in ranked if c.classification.role == ROLE_DOCK_BRIDGE), None)
    if handheld.architecture == ARCH_SH3:
        if dock_bridge is not None:
            return [Finding(LEVEL_OK, f"{handheld.name} is linked through the dock's USB jack",
                            f"The bridge retrofitted into the dock ({dock_bridge.device.label}) carries the "
                            f"handheld's RS-232 lines to {dock_bridge.path}; the SH-3 itself never sees USB.")]
        return [Finding(LEVEL_INFO, f"{handheld.name}: the dock's USB-B jack is inert for this model",
                        "The SH7709A CPU and HD64461 companion chip have no USB device controller, so the "
                        "connector's USB pins carry nothing. Connect the dock's DB-9 (or the sync cable) to a "
                        f"USB-serial adapter, or retrofit a bridge into the dock — see {DOCS_PATH}.")]
    if handheld.architecture == ARCH_ARM:
        return [Finding(LEVEL_INFO, f"{handheld.name} can enumerate on the dock's USB jack",
                        "Its SA-1110 has a USB device controller and Windows CE 3.0 presents a USB Sync "
                        "function, but macOS ships no driver for it; jornada-link's native USB link is on the "
                        f"roadmap ({DOCS_PATH}). Use the serial path meanwhile.")]
    return []


def _sync_device_findings(items: Tuple[ClassifiedDevice, ...],
                          handheld: Optional[HandheldModel]) -> List[Finding]:
    findings: List[Finding] = []
    for item in items:
        if item.role != ROLE_WINCE_USB_SYNC:
            continue
        name = item.classification.profile.name if item.classification else item.device.label
        findings.append(Finding(
            LEVEL_INFO, f"Windows CE USB Sync device present ({item.device.vid_pid}, {name})",
            "A handheld speaking USB natively. macOS has no driver for the CE USB Sync function; "
            f"jornada-link's userspace USB link for it is on the roadmap ({DOCS_PATH})."))
        if handheld is not None and handheld.architecture == ARCH_SH3:
            findings.append(Finding(
                LEVEL_WARN, "That USB Sync device is not the SH-3 Jornada",
                f"{handheld.name} has no USB controller, so another Windows CE device is attached."))
    return findings


def _adapter_findings(items: Tuple[ClassifiedDevice, ...], ranked: Tuple[Candidate, ...],
                      recommended: Optional[str]) -> List[Finding]:
    findings: List[Finding] = []
    for item in items:
        if item.classification is None or item.role == ROLE_WINCE_USB_SYNC:
            continue
        profile = item.classification.profile
        nodes = item.device.serial_nodes
        if not nodes:
            hint = ("macOS should have created one by itself; try another port or cable."
                    if profile.macos_driver == DRIVER_BUILT_IN
                    else f"install the vendor driver ({profile.driver_name}) and reconnect the adapter.")
            findings.append(Finding(LEVEL_ERROR, f"{profile.name} has no serial node",
                                    f"{item.device.label} ({item.device.vid_pid}) is attached but no /dev/cu.* "
                                    f"node exists for it — {hint}"))
            continue
        if len(nodes) > 1:
            drivers = ", ".join(f"{n.path} ({n.driver})" for n in nodes)
            findings.append(Finding(LEVEL_INFO, f"{profile.name} exposes {len(nodes)} serial nodes",
                                    f"Two drivers are bound to the same adapter: {drivers}. "
                                    f"The link uses {recommended or nodes[0].path}."))
        for node in nodes:
            if not node.writable:
                findings.append(Finding(LEVEL_WARN, f"{node.path} is root-only",
                                        f"The node created by {node.driver} is not openable by this user. The root "
                                        "pppd can still use it, but `jornada probe` and the pre-connect flush cannot, "
                                        "so the link prefers another node of the same adapter when there is one."))
    return findings


def diagnose(devices: Tuple[UsbDevice, ...], pinned: Optional[str] = None,
             handheld: Optional[HandheldModel] = None) -> Diagnosis:
    """The complete picture for a snapshot of the USB bus (pure; no I/O)."""
    items = classify_devices(devices)
    ranked = rank_candidates(devices, pinned)
    recommended = ranked[0].path if ranked else None
    findings: List[Finding] = []

    if pinned is not None and not any(c.path == pinned for c in ranked):
        findings.append(Finding(LEVEL_WARN, f"Pinned serial port {pinned} is not present",
                                "The pin in ~/.jornada-link/serial does not match any attached adapter; "
                                "the best attached adapter is used instead. Re-pin or unpin it."))
    if not ranked and not any(i.role == ROLE_WINCE_USB_SYNC for i in items):
        findings.append(Finding(LEVEL_ERROR, "No USB-serial adapter found",
                                "Plug the adapter into the Jornada's sync cable or the dock's DB-9 port. FTDI "
                                "adapters need no driver on macOS; Prolific and WCH chips usually need the "
                                "vendor's DriverKit extension."))
    findings.extend(_sync_device_findings(items, handheld))
    findings.extend(_dock_findings(handheld, ranked))
    findings.extend(_adapter_findings(items, ranked, recommended))
    if recommended is not None:
        best = ranked[0]
        via = "pinned" if pinned == best.path else "best available"
        access = "openable by this user" if best.writable else "root-only (unpin to let the doctor choose)"
        findings.append(Finding(LEVEL_OK, f"Serial link port: {best.path} ({via})",
                                f"{best.classification.profile.name}, driver {best.driver}, {access}."))
    return Diagnosis(devices=items, candidates=ranked, recommended=recommended, pinned=pinned,
                     handheld=handheld, findings=tuple(findings))

File: macapp/Sources/JornadaCore/UsbDoctor.swift
import Foundation

/// Which USB device should carry the link, and why or why not — Swift twin of
/// `jornada/usb_doctor.py`. Pure: takes a snapshot of the bus, the pinned port
/// and the known handheld; produces ranked candidates and human findings.
public enum UsbFindingLevel: String, Sendable, Comparable {
    case ok, info, warn, error

    private var rank: Int { [Self.ok: 0, .info: 1, .warn: 2, .error: 3][self] ?? 0 }
    public static func < (lhs: Self, rhs: Self) -> Bool { lhs.rank < rhs.rank }
}

public struct UsbFinding: Sendable, Equatable, Identifiable {
    public let level: UsbFindingLevel
    public let title: String
    public let detail: String
    public var id: String { "\(level.rawValue):\(title)" }
}

public struct UsbClassifiedDevice: Sendable, Identifiable {
    public let device: UsbDevice
    public let classification: UsbClassification?
    public var id: String { device.id }
    public var role: UsbRole? { classification?.role }
}

public struct UsbCandidate: Sendable, Identifiable {
    public let path: String
    public let driver: String
    public let writable: Bool
    public let device: UsbDevice
    public let classification: UsbClassification
    public let score: Int
    public var id: String { path }
}

public struct UsbDiagnosis: Sendable {
    public let devices: [UsbClassifiedDevice]
    public let candidates: [UsbCandidate]
    public let recommended: String?
    public let pinned: String?
    public let handheld: HandheldModel?
    public let findings: [UsbFinding]

    public var worstLevel: UsbFindingLevel { findings.map(\.level).max() ?? .ok }
}

public enum UsbDoctor {
    public static let docsPath = "docs/usb-link.md"

    public static func classify(_ devices: [UsbDevice]) -> [UsbClassifiedDevice] {
        UsbRegistry.linkCandidates(devices).map { device in
            UsbClassifiedDevice(device: device, classification: UsbProfiles.classify(
                vendorId: device.vendorId, productId: device.productId, product: device.product,
                serial: device.serial, interfaceClasses: device.interfaceClasses))
        }
    }

    static func score(role: UsbRole, driver: String, writable: Bool, pinned: Bool) -> Int {
        var score = role == .dockBridge ? 400 : (role == .serialBridge ? 200 : 0)
        if driver.hasPrefix("com.apple.") { score += 100 }
        score += writable ? 50 : -500
        if pinned { score += 10_000 }
        return score
    }

    /// Every serial node of every recognised bridge, best first.
    public static func rankCandidates(_ devices: [UsbDevice], pinned: String? = nil) -> [UsbCandidate] {
        var found: [UsbCandidate] = []
        for item in classify(devices) {
            guard let classification = item.classification, classification.role != .winceUsbSync else { continue }
            for node in item.device.serialNodes {
                found.append(UsbCandidate(
                    path: node.path, driver: node.driver, writable: node.writable, device: item.device,
                    classification: classification,
                    score: score(role: classification.role, driver: node.driver, writable: node.writable,
                                 pinned: pinned == node.path)))
            }
        }
        return found.sorted { ($0.score, $1.path) > ($1.score, $0.path) }
    }

    public static func recommendedSerialPath(_ devices: [UsbDevice], pinned: String? = nil) -> String? {
        rankCandidates(devices, pinned: pinned).first?.path
    }

    static func dockFindings(_ handheld: HandheldModel?, _ ranked: [UsbCandidate]) -> [UsbFinding] {
        guard let handheld else { return [] }
        let dockBridge = ranked.first { $0.classification.role == .dockBridge }
        switch handheld.architecture {
        case .sh3:
            if let bridge = dockBridge {
                return [UsbFinding(level: .ok, title: "\(handheld.name) is linked through the dock's USB jack",
                                   detail: "The bridge retrofitted into the dock (\(bridge.device.label)) carries the "
                                       + "handheld's RS-232 lines to \(bridge.path); the SH-3 itself never sees USB.")]
            }
            return [UsbFinding(level: .info, title: "\(handheld.name): the dock's USB-B jack is inert for this model",
                               detail: "The SH7709A CPU and HD64461 companion chip have no USB device controller, so the "
                                   + "connector's USB pins carry nothing. Connect the dock's DB-9 (or the sync cable) to a "
                                   + "USB-serial adapter, or retrofit a bridge into the dock — see \(docsPath).")]
        case .arm:
            return [UsbFinding(level: .info, title: "\(handheld.name) can enumerate on the dock's USB jack",
                               detail: "Its SA-1110 has a USB device controller and Windows CE 3.0 presents a USB Sync "
                                   + "function, but macOS ships no driver for it; jornada-link's native USB link is on the "
                                   + "roadmap (\(docsPath)). Use the serial path meanwhile.")]
        }
    }

    static func syncDeviceFindings(_ items: [UsbClassifiedDevice], _ handheld: HandheldModel?) -> [UsbFinding] {
        var findings: [UsbFinding] = []
        for item in items where item.role == .winceUsbSync {
            let name = item.classification?.profile.name ?? item.device.label
            findings.append(UsbFinding(
                level: .info, title: "Windows CE USB Sync device present (\(item.device.vidPid), \(name))",
                detail: "A handheld speaking USB natively. macOS has no driver for the CE USB Sync function; "
                    + "jornada-link's userspace USB link for it is on the roadmap (\(docsPath))."))
            if let handheld, handheld.architecture == .sh3 {
                findings.append(UsbFinding(level: .warn, title: "That USB Sync device is not the SH-3 Jornada",
                                           detail: "\(handheld.name) has no USB controller, so another Windows CE device is attached."))
            }
        }
        return findings
    }

    static func adapterFindings(_ items: [UsbClassifiedDevice], recommended: String?) -> [UsbFinding] {
        var findings: [UsbFinding] = []
        for item in items {
            guard let classification = item.classification, classification.role != .winceUsbSync else { continue }
            let profile = classification.profile
            let nodes = item.device.serialNodes
            if nodes.isEmpty {
                let hint = profile.macosDriver == .builtIn
                    ? "macOS should have created one by itself; try another port or cable."
                    : "install the vendor driver (\(profile.driverName)) and reconnect the adapter."
                findings.append(UsbFinding(level: .error, title: "\(profile.name) has no serial node",
                                           detail: "\(item.device.label) (\(item.device.vidPid)) is attached but no /dev/cu.* "
                                               + "node exists for it — \(hint)"))
                continue
            }
            if nodes.count > 1 {
                let drivers = nodes.map { "\($0.path) (\($0.driver))" }.joined(separator: ", ")
                findings.append(UsbFinding(level: .info, title: "\(profile.name) exposes \(nodes.count) serial nodes",
                                           detail: "Two drivers are bound to the same adapter: \(drivers). "
                                               + "The link uses \(recommended ?? nodes[0].path)."))
            }
            for node in nodes where !node.writable {
                findings.append(UsbFinding(level: .warn, title: "\(node.path) is root-only",
                                           detail: "The node created by \(node.driver) is not openable by this user. The root "
                                               + "pppd can still use it, but `jornada probe` and the pre-connect flush cannot, "
                                               + "so the link prefers another node of the same adapter when there is one."))
            }
        }
        return findings
    }

    /// The complete picture for a snapshot of the USB bus.
    public static func diagnose(_ devices: [UsbDevice], pinned: String? = nil,
                                handheld: HandheldModel? = nil) -> UsbDiagnosis {
        let items = classify(devices)
        let ranked = rankCandidates(devices, pinned: pinned)
        let recommended = ranked.first?.path
        var findings: [UsbFinding] = []

        if let pinned, !ranked.contains(where: { $0.path == pinned }) {
            findings.append(UsbFinding(level: .warn, title: "Pinned serial port \(pinned) is not present",
                                       detail: "The pin in ~/.jornada-link/serial does not match any attached adapter; "
                                           + "the best attached adapter is used instead. Re-pin or unpin it."))
        }
        if ranked.isEmpty, !items.contains(where: { $0.role == .winceUsbSync }) {
            findings.append(UsbFinding(level: .error, title: "No USB-serial adapter found",
                                       detail: "Plug the adapter into the Jornada's sync cable or the dock's DB-9 port. FTDI "
                                           + "adapters need no driver on macOS; Prolific and WCH chips usually need the "
                                           + "vendor's DriverKit extension."))
        }
        findings += syncDeviceFindings(items, handheld)
        findings += dockFindings(handheld, ranked)
        findings += adapterFindings(items, recommended: recommended)
        if let best = ranked.first {
            let via = pinned == best.path ? "pinned" : "best available"
            let access = best.writable ? "openable by this user" : "root-only (unpin to let the doctor choose)"
            findings.append(UsbFinding(level: .ok, title: "Serial link port: \(best.path) (\(via))",
                                       detail: "\(best.classification.profile.name), driver \(best.driver), \(access)."))
        }
        return UsbDiagnosis(devices: items, candidates: ranked, recommended: recommended, pinned: pinned,
                            handheld: handheld, findings: findings)
    }

    // MARK: - The pin file (~/.jornada-link/serial), shared with the CLI and bin/jornada-ppp

    public static func pinFile(directory: URL = PppController.stateDirectory()) -> URL {
        directory.appendingPathComponent("serial")
    }

    public static func readPin(directory: URL = PppController.stateDirectory()) -> String? {
        guard let text = try? String(contentsOf: pinFile(directory: directory), encoding: .utf8) else { return nil }
        let first = text.split(whereSeparator: \.isNewline).first.map { $0.trimmingCharacters(in: .whitespaces) } ?? ""
        return PppController.isValidSerialPath(first) ? first : nil
    }

    public static func writePin(_ path: String, directory: URL = PppController.stateDirectory()) throws {
        guard PppController.isValidSerialPath(path) else {
            throw PppController.ControlError.invalidSerialPath(path)
        }
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        try (path + "\n").write(to: pinFile(directory: directory), atomically: true, encoding: .utf8)
    }

    public static func clearPin(directory: URL = PppController.stateDirectory()) {
        try? FileManager.default.removeItem(at: pinFile(directory: directory))
    }
}
